src/MaxwellSolvers/FDTD.h

```diff
@@ -0,0 +1,403 @@
+#ifndef IPPL_FDTD_H
+#define IPPL_FDTD_H
+#include <cstddef>
+using std::size_t;
+#include "Types/Vector.h"
+
+#include "FieldLayout/FieldLayout.h"
+#include "MaxwellSolvers/AbsorbingBC.h"
+#include "MaxwellSolvers/Maxwell.h"
+#include "Meshes/UniformCartesian.h"
+#include "Particle/ParticleBase.h"
+
+constexpr double sq(double x) {
+    return x * x;
+}
+constexpr float sq(float x) {
+    return x * x;
+}
+#define assert_isreal(X) assert(!Kokkos::isnan(X) && !Kokkos::isinf(X))
+namespace ippl {
+    enum fdtd_bc {
+        periodic,
+        absorbing
+    };
+
+    template <typename EMField, typename FourField, fdtd_bc boundary_conditions = periodic>
+    class StandardFDTDSolver : Maxwell<EMField, FourField> {
+    public:
+        constexpr static unsigned Dim = EMField::dim;
+        using scalar                  = typename EMField::value_type::value_type;
+        using Vector_t                = Vector<typename EMField::value_type::value_type, Dim>;
+        using FourVector_t            = typename FourField::value_type;
+        StandardFDTDSolver(FourField& source, EMField& E, EMField& B) {
+            Maxwell<EMField, FourField>::setSource(source);
+            Maxwell<EMField, FourField>::setEMFields(E, B);
+            initialize();
+        }
+        virtual void solve() override {
+            step();
+            timeShift();
+            evaluate_EB();
+        }
+        void setPeriodicBoundaryConditions() {
+            periodic_bc = true;
+            typename FourField::BConds_t vector_bcs;
+            auto bcsetter_single = [&vector_bcs]<size_t Idx>(const std::index_sequence<Idx>&) {
+                vector_bcs[Idx] = std::make_shared<ippl::PeriodicFace<FourField>>(Idx);
+                return 0;
+            };
+            auto bcsetter = [bcsetter_single]<size_t... Idx>(const std::index_sequence<Idx...>&) {
+                int x = (bcsetter_single(std::index_sequence<Idx>{}) ^ ...);
+                (void)x;
+            };
+            bcsetter(std::make_index_sequence<Dim * 2>{});
+            A_n.setFieldBC(vector_bcs);
+            A_np1.setFieldBC(vector_bcs);
+            A_nm1.setFieldBC(vector_bcs);
+        }
+
+        FourField A_n;
+        FourField A_np1;
+        FourField A_nm1;
+        scalar dt;
```

Comment:
Previous review comment still applies: Are these attributes supposed to be public?
It would also make the code more readable to have them at the end of the code block (like in the first public block you have the attributes at the end before the private).

Reply:
These are protected variables in FDTDSolverBase now.